File: 500.600/524.LongestWordInDictionaryThroughDeleting.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

class LongestWordInDictionaryThroughDeleting
{
public:
    string findLongestWord(string s, const vector<string> &dictionary)
    {
        int cur = -1;
        int n = dictionary.size();
        for (int i = 0; i < n; ++i)
        {
            if (isSubstr(dictionary[i], s))
            {
                if (cur != -1)
                {
                    if (dictionary[cur].size() < dictionary[i].size())
                    {
                        cur = i;
                    }
                    else if (dictionary[cur].size() == dictionary[i].size() && dictionary[cur] > dictionary[i])
                    {
                        cur = i;
                    }
                }
                else
                {
                    cur = i;
                }
            }
        }
        if (cur == -1)
        {
            return "";
        }
        return dictionary[cur];
    }

private:
    bool isSubstr(const string &t, const string &s)
    {
        int pt = 0, ps = 0;
        int nt = t.size(), ns = s.size();
        if (nt > ns)
        {
            return false;
        }
        while (pt < nt && ps < ns)
        {
            if (t[pt] == s[ps])
            {
                ++pt;
            }
            ++ps;
        }
        return pt == nt;
    }

public:
    // 先排序再匹配
    string findLongestWord1(string s, vector<string> dictionary)
    {
        sort(dictionary.begin(), dictionary.end(), [](const string &a, const string &b)
             { return a.size() > b.size() || (a.size() == b.size() && a < b); });
        for (const string &t : dictionary)
        {
            if (isSubstr(t, s))
            {
                return t;
            }
        }
        return "";
    }
};

int main()
{
    LongestWordInDictionaryThroughDeleting cls;
    cout << cls.findLongestWord1("abpcplea", vector<string>{"ale", "apple", "monkey", "plea"}) << " -> apple" << endl;
    cout << cls.findLongestWord1("abpcplea", vector<string>{"a", "b", "c"}) << " -> a" << endl;
    return 0;
}

File: go/500.600/524.LongestWordInDictionaryThroughDeleting.go
package main

import (
	"fmt"
	"sort"
)

func findLongestWord(s string, dictionary []string) string {
	cur := -1
	for i, t := range dictionary {
		if isSubstr(&t, &s) {
			if cur == -1 {
				cur = i
			} else if len(t) > len(dictionary[cur]) {
				cur = i
			} else if len(t) == len(dictionary[cur]) && t < dictionary[cur] {
				cur = i
			}
		}
	}
	if cur == -1 {
		return ""
	}
	return dictionary[cur]
}

func isSubstr(t, s *string) bool {
	nt, ns := len(*t), len(*s)
	if nt > ns {
		return false
	}
	pt, ps := 0, 0
	for pt < nt && ps < ns {
		if (*t)[pt] == (*s)[ps] {
			pt++
		}
		ps++
	}
	return pt == nt
}

func findLongestWord1(s string, dictionary []string) string {
	sort.Slice(dictionary, func(i, j int) bool {
		return len(dictionary[i]) > len(dictionary[j]) || (len(dictionary[i]) == len(dictionary[j]) && dictionary[i] < dictionary[j])
	})
	for _, t := range dictionary {
		if isSubstr(&t, &s) {
			return t
		}
	}
	return ""
}

func main() {
	fmt.Println(findLongestWord1("abpcplea", []string{"ale", "apple", "monkey", "plea"}), "apple")
	fmt.Println(findLongestWord1("abpcplea", []string{"a", "b", "c"}), "a")
}
